### PR TITLE
Improved defense play

### LINKED ISSUE #2385: CreaseDefensePlayFSM: Handle Multiple Threats and Multiple Crease Defenders
### Description of the task
The CreaseDefensePlayFSM is currently limited to three robots and only handles one threat. It also does not spread robots out neatly, for example two robots would take the centre and left positions with respect to the threat. An advanced CreaseDefensePlayFSM would be able to handle any number of threats with any number of available friendly robots and position robots optimally. Additionally, it would be able to dispatch robots to intercept incoming balls.

### Acceptance criteria

- [ ] Design and implement such as CreaseDefensePlayFSM
- [ ] Testing

### LINKED ISSUE #2440: Add Adaptive Offense/Defense to Playing State
### Description of the task
Depending on which team has possession, number of robots on each team, etc., we should allocate robots to different tasks when the game is in the playing state.

### Acceptance criteria

- [ ] Look for TODO with this issue
- [ ] Add smart logic to dynamic allocate robots to shadowing, attacking, etc. depending on the state of plaay

### LINKED ISSUE #2862: Group together enemy offense clusters when distributing defenders
### Description of the task

Teams like Tigers have implemented a grouping system in their defense strategy, where enemy robots with similar positioning/angle to the goal are grouped together and treated as a single offensive cluster ([TDP, see 2.3 Defensive](https://ssl.robocup.org/wp-content/uploads/2019/01/2017_ETDP_TIGERs_Mannheim.pdf)). This concept helps prevent the scenario where the enemy team "piles" the majority of it's robots into a corner or small zone of the field, and puts one attacker/pass receiver further away. Our current defense would assign most defenders to defend against the pile, leaving the lone attacker mostly undefended and open to score a clear goal.
 
### Acceptance criteria

- [ ] Implement enemy grouping in our defense play FSMs
- [ ] Write tests for the "pile" scenario to test whether the grouping system improves our number of goals allowed

### PR DESCRIPTION
**Pull request by williamckha**
### Description

Improved defense play + smarter transitioning between offensive and defensive gameplay

- Added defense play FSM and new logic for determining tactic assignments + defender positioning
- Usage of pass defender tactic (#2817) in new defense play
- Crease defenders spread out to cover multiple threats and align more neatly to target threat
- Added possession tracker for determining team possession of the ball given the state of the game over time
- Treat enemy robots with similar positioning/angle to the goal as a single offensive cluster
- Converted offense play to FSM

### Testing Done

Run AI vs AI for offensive and defensive gameplay
Run `defense_play_test.py` for yellow offense vs blue defense

### Resolved Issues

Resolves #2385, resolves #2440, resolves #2862

### Length Justification and Key Files to Review

Most of the new defense is under the `evaluation` and `defense` folders

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/evaluation/possession.cpp [hunk 1]
```diff
@@ -2,10 +2,125 @@
...
 #include "software/world/team.h"
 
+PossessionTracker::PossessionTracker() {}
+
+TeamPossession PossessionTracker::getTeamWithPossession(const Team &friendly_team,
+                                                        const Team &enemy_team,
+                                                        const Ball &ball,
+                                                        const Field &field)
+{
+    // Based on CMDragons TDP 2015
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.cpp]: nit: could you link their TDP here as well?

File: src/software/ai/evaluation/possession.h [hunk 2]
```diff
@@ -1,7 +1,67 @@
...
+    // to be considered far away from the ball
+    static constexpr double DISTANCE_FAR_TOLERANCE_METERS = 0.5;
+
+    // Min time in seconds that robot must stay close to ball for
+    // it to be considered near
+    static constexpr double TIME_NEAR_THRESHOLD_S = 0.1;
+
+    // Min time in seconds that robot must stay away from ball for
+    // it to be considered far away
+    static constexpr double TIME_FAR_THRESHOLD_S = 1.5;
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.h]: If you think these values will be tuned further and we might want to change them during a game, it is a good idea to define them as part of our configurable parameters in `parameters.proto` and dependency injected through the constructor from the class that creates the `PossessionTracker`.

If you decide to keep the constants, you can probably convert the time constants to `Duration`s (so you dont have to do the conversion in `getTeamWithPossession`.

File: src/software/ai/evaluation/possession.h [hunk 3]
```diff
@@ -1,7 +1,67 @@
...
+   private:
+    Timestamp last_timestamp;
+    Duration time_near_friendly;
+    Duration time_near_enemy;
+    Duration time_far_friendly;
+    Duration time_far_enemy;
+    TeamPossession possession;
+
+    void updateTimes(const Team &friendly_team, const Team &enemy_team, const Ball &ball,
+                     const Field &field);
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.h]: nit: Missing docs

File: src/software/ai/evaluation/possession.cpp [hunk 4]
```diff
@@ -2,10 +2,125 @@
 
 #include "shared/constants.h"
 #include "software/ai/evaluation/intercept.h"
+#include "software/geom/algorithms/distance.h"
 #include "software/world/ball.h"
 #include "software/world/field.h"
 #include "software/world/team.h"
 
+PossessionTracker::PossessionTracker() {}
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.cpp]: It's a good idea to initialize the fields to clarify what the starting conditions are (e.g. It's a bit ambiguous what the `possession` is on initialization, specially since `getTeamWithPossession` doesn't guarantee that it sets the `possession` variable.

File: src/software/ai/evaluation/possession.cpp [hunk 5]
```diff
@@ -2,10 +2,125 @@
...
+             (time_near_enemy > TIME_NEAR_THRESHOLD))
+    {
+        possession = TeamPossession::ENEMY_TEAM;
+    }
+    else if ((time_near_friendly > TIME_NEAR_THRESHOLD) &&
+             (time_near_enemy > TIME_NEAR_THRESHOLD))
+    {
+        // Both teams are considered to have presence over the ball.
+        // If the ball is on our side of the field, or no enemy robots are
+        // on our side of the field, consider enemy team as having possession.
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.cpp]: This is an interesting approach. I gotta get some more testing with it to see how it behaves. 
There are some scenarios which I'm unsure about. E.g. Imagine we have the ball in our dribbler, in our own half, if an enemy robot comes near our robot (for atleast 0.1sec), then we say that the enemy has the possession. 

I wonder if we should be incorporating the breakbeam as well. i.e. If the breakbeam is triggered, we know the ball is in our possession (though you could also have a scenario where we're face to face with an enemy robot and they also have the ball near/in their dribbler...). This is already something we kind of do in Sensor Fusion:

https://github.com/UBC-Thunderbots/Software/blob/97acde5458f2d79b36d99f7143139a3ea6a93041/src/software/sensor_fusion/sensor_fusion.cpp#L172-L178

File: src/software/ai/evaluation/possession.cpp [hunk 6]
```diff
@@ -2,10 +2,125 @@
...
 #include "software/world/ball.h"
 #include "software/world/field.h"
 #include "software/world/team.h"
 
+PossessionTracker::PossessionTracker() {}
+
+TeamPossession PossessionTracker::getTeamWithPossession(const Team &friendly_team,
+                                                        const Team &enemy_team,
+                                                        const Ball &ball,
+                                                        const Field &field)
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.cpp]: It seems like we already do some possession estimation in Sensor Fusion as well (which is stored in the `World::team_with_possesion_`. Though it seems like that is unused). @itsarune Is this what we used previously for possession estimation which resulted in oscillations between plays?
https://github.com/UBC-Thunderbots/Software/blob/97acde5458f2d79b36d99f7143139a3ea6a93041/src/software/sensor_fusion/sensor_fusion.cpp#L290-L306

File: src/software/ai/evaluation/possession.h [hunk 7]
```diff
@@ -1,7 +1,67 @@
...
+#include "software/time/duration.h"
+#include "software/time/timestamp.h"
 #include "software/world/world.h"
 
+// Indicates which team has possession of the ball
+enum TeamPossession
+{
+    FRIENDLY_TEAM,
+    ENEMY_TEAM
+};
```
Comment (nimazareian) [inline src/software/ai/evaluation/possession.h]: Curious if you've considered using the existing `TeamType` enum?
https://github.com/UBC-Thunderbots/Software/blob/82d876b16aa75666173bd28b2744782b076d5b2c/src/software/world/team.h#L271-L275

File: src/software/ai/evaluation/possession.cpp
Comment (nimazareian) [inline src/software/ai/evaluation/possession.cpp:1]: Do you think you could write some basic tests for this class? They should probably be testing cases where the possession is obvious, and you can leave edge cases where it's up for interpretation/implementation (i.e. The tests shouldn't be implementation specific)

Similarly for the functions in `defender_assignment.cpp`?

File: src/software/ai/evaluation/defender_assignment.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+// Multiplier to ensure that goal lanes are scored higher (in threat rating)
+// than passing lanes
+static constexpr double GOAL_LANE_THREAT_MULTIPLIER = 3;
+
+// Max percent difference between two goal lanes' angles to goal for them to be
+// considered a dense lane cluster
+static constexpr double GOAL_LANE_DENSITY_THRESHOLD = 0.3;
+
+// Bonus added to coverage rating if the goal lane is not part of a dense cluster
+static constexpr double GOAL_LANE_NONDENSE_BONUS = 0.5;
```
Comment (nimazareian) [inline src/software/ai/evaluation/defender_assignment.h]: some/all of these constants should also be added to `parameters.proto` so they're configurable. Could be a new type added to `DefensePlayConfig`

File: src/software/ai/hl/stp/play/defense/defense_play.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
+#include "software/ai/hl/stp/play/defense/defense_play.h"
+
+#include "proto/parameters.pb.h"
+#include "shared/constants.h"
+#include "software/logger/logger.h"
+#include "software/util/generic_factory/generic_factory.h"
+
+DefensePlay::DefensePlay(TbotsProto::AiConfig config)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/defense/defense_play.cpp]: nit
```suggestion
DefensePlay::DefensePlay(const TbotsProto::AiConfig& config)
```

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        auto target = crease_defender_assignments.at(i).target;
+
+        // Determine the number of crease defenders already assigned to the target
+        auto defenders_with_target_count = std::count_if(
+            crease_defender_assignments.begin(), crease_defender_assignments.begin() + i,
+            [&target](const auto& assignment) { return assignment.target == target; });
+
+        // Pick alignment based on how many crease defenders are already assigned to the
+        // target
+        TbotsProto::CreaseDefenderAlignment alignment;
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: So is `CreaseDefensePlay` not used anymore?

**Review by nimazareian — COMMENTED** [review]
Tried out the updated plays in Thunderscope and it looks great! Our defense is definitely much stronger and better positioned than what we had before! Great job :100: 

I noticed that sometimes both robots would stop going for the ball when the ball was stationary near them. Do you know why this could be happening? Is it related to the possession tracker? Here's a scenario where both team stopped moving
image


I think this is unrelated to the changes in this PR, but we should _definitely_ try to fix it before robocup (tagging @itsarune and @sauravbanna as well). I was running AI vs AI on this branch and friendly robot 2 was waiting for a pass in the friendly half, which is very risky and doesn't really benefit us. I think we should consider modifying the cost functions to avoid this from happening!
image

**Comment by williamckha** [comment]
> I noticed that sometimes both robots would stop going for the ball when the ball was stationary near them. Do you know why this could be happening? Is it related to the possession tracker? Here's a scenario where both team stopped moving

I'll try to reproduce this scenario and take a closer look. Does this happen only when the robots/ball are near the crease? If so #2807 might be related

File: src/software/ai/evaluation/possession.h (hunk 7 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.h]: I was experimenting with having NO_TEAM to indicate no team has clear possession over the ball and different degrees of possession (e.g. AGGRESSIVE_ENEMY_TEAM), figured it would be easier to make changes if we end up returning to those ideas

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 10 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: Yeah its not used anymore.

File: src/software/ai/evaluation/possession.cpp (hunk 5 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.cpp]: Good idea re: breakbeam. For some scenarios the breakbeam data should be useful. E.g. If the breakbeam is triggered and we know that the enemy team is far away from ball, then we can probably be confident that we have possession.

Maybe we should introduce some different types of possession that trigger more cautious strategies in OffensePlay. E.g. in the scenario you described, we could return something like `TeamPossession::FRIENDLY_GUARDING` to indicate that possession is being contested and that the friendly team should try to protect the ball. This could initiate a more offensive defense with a ball retrieval/dribble away tactic

Comment (nimazareian) [reply src/software/ai/evaluation/possession.cpp]: That's a great idea! It'd be really interesting having different levels of offense/defense depending on the scenario and the placement of the robots. 

> E.g. If the breakbeam is triggered and we know that the enemy team is far away from ball, then we can probably be confident that we have possession.

That's right, though we do have to be slightly careful, since there could be scenarios where we have the ball in our dribbler, but the enemy robot is right in our face (potentially with the ball touching their dribbler as well) trying to steel the ball from us. Switching to an offensive play in that scenario is risky.

File: src/software/ai/evaluation/possession.h (hunk 7 above)
Comment (nimazareian) [reply src/software/ai/evaluation/possession.h]: Ok that makes more sense now

**Comment by nimazareian** [comment]
> I'll try to reproduce this scenario and take a closer look. Does this happen only when the robots/ball are near the crease? If so #2807 might be related

Yes I believe it only happened with the ball being near the crease defenders, though it was away from the defense area (similar to the picture I posted above).

File: src/software/ai/hl/stp/play/defense/defense_play.cpp (hunk 9 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/defense/defense_play.cpp]: done

File: src/software/ai/evaluation/possession.h (hunk 3 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.h]: done

File: src/software/ai/evaluation/possession.h (hunk 2 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.h]: done

File: src/software/ai/evaluation/possession.cpp (hunk 1 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.cpp]: done

File: src/software/ai/evaluation/possession.h (hunk 2 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.h]: done

File: src/software/ai/evaluation/possession.cpp (hunk 4 above)
Comment (williamckha) [reply src/software/ai/evaluation/possession.cpp]: done

File: src/software/ai/evaluation/defender_assignment.h (hunk 8 above)
Comment (williamckha) [reply src/software/ai/evaluation/defender_assignment.h]: done

**Review by nimazareian — COMMENTED** [review]
Changes look good. Just pending some test cases and a possible update on robots stopping to go for the ball.

**Comment by williamckha** [comment]
I managed to recreate the scenario and it looks like all the bots have the correct tactics and their FSMs are in the correct states. The blue bots are assigned `CreaseDefenderTactic` and the yellow attacker that's stuck near the ball is assigned `AttackerTactic` in the state `PivotKickFSM.DribbleFSM.GetPossession`. Both teams stop moving probably due to the issue described in #2807 because from the yellow team POV, the ball is in the dead zone created by the inflated defense area obstacle


| Blue Fullsystem | Yellow Fullsystem |
| --- | --- |
| Screenshot 2023-05-27 102307 | Screenshot 2023-05-27 102953 |

Working on writing tests + trying out some different ideas with possession tracker, I'll notify when done

File: src/software/ai/evaluation/defender_assignment.cpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            DefenderAssignment{PASS_DEFENDER, lane.midPoint(), threat_rating});
+    }
+
+    // Construct goal lanes.
+    // Using full list of threats (not filtered threats) since we need to
+    // find potential goal lane from every enemy on the field
+    for (unsigned int i = 0; i < relevant_threats.size(); i++)
+    {
+        auto threat_position =
+            (i == 0) ? ball.position() : relevant_threats.at(i).robot.position();
```
Comment (itsarune) [inline src/software/ai/evaluation/defender_assignment.cpp]: we typically avoid ? syntax for clarity. prefer if/else here

File: src/software/ai/evaluation/defender_assignment.cpp [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    auto grouped_goal_lanes =
+        groupGoalLanesByDensity(goal_lanes, config.goal_lane_density_threshold());
+
+    // Determine crease defender assignments based on goal lanes
+    for (const auto &goal_lanes_group : grouped_goal_lanes)
+    {
+        // We include a non-dense bonus when rating crease defender assignment
+        // if the goal lane is not part of a dense cluster
+        double nondense_bonus =
+            (goal_lanes_group.size() == 1) ? config.goal_lane_nondense_bonus() : 0;
```
Comment (itsarune) [inline src/software/ai/evaluation/defender_assignment.cpp]: prefer if/else here for clarity

File: src/software/ai/evaluation/possession.cpp (hunk 6 above)
Comment (itsarune) [reply src/software/ai/evaluation/possession.cpp]: this approach solved oscillation between plays. For context, I believe we used to keep track of x number of frames and used previous frame history to figure out if we still have possession or not.

File: src/software/ai/evaluation/possession.cpp [hunk 13]
```diff
@@ -2,10 +2,135 @@
...
+        possession = TeamPossession::FRIENDLY_TEAM;
+    }
+    else if ((time_near_friendly < time_near_threshold) &&
+             (time_near_enemy > time_near_threshold))
+    {
+        possession = TeamPossession::ENEMY_TEAM;
+    }
+    else if ((time_near_friendly > time_near_threshold) &&
+             (time_near_enemy > time_near_threshold))
+    {
```
Comment (itsarune) [inline src/software/ai/evaluation/possession.cpp]: we haven't tried a time-based possession approach. I wonder if this may be a better way of doing it.

File: src/software/math/math_functions.h [hunk 14]
```diff
@@ -93,3 +93,16 @@ T normalizeValueToRange(T value, const T& value_min, const T& value_max,
...
+
+/**
+ * Calculates the percent difference between two values.
+ *
+ * @param a The first value
+ * @param b The second value
+ *
+ * @return The percent difference between the two values
+ */
+inline double percent_difference(double a, double b)
```
Comment (itsarune) [inline src/software/math/math_functions.h]: add a simple test for this

File: src/software/ai/evaluation/possession.cpp
Comment (williamckha) [reply src/software/ai/evaluation/possession.cpp:1]: done

**Comment by nimazareian** [comment]
> I managed to recreate the scenario and it looks like all the bots have the correct tactics and their FSMs are in the correct states. The blue bots are assigned `CreaseDefenderTactic` and the yellow attacker that's stuck near the ball is assigned `AttackerTactic` in the state `PivotKickFSM.DribbleFSM.GetPossession`. Both teams stop moving probably due to the issue described in #2807 because from the yellow team POV, the ball is in the dead zone created by the inflated defense area obstacle
> 
> Blue Fullsystem	Yellow Fullsystem
> Screenshot 2023-05-27 102307	Screenshot 2023-05-27 102953
> Working on writing tests + trying out some different ideas with possession tracker, I'll notify when done

I see. Though I feel like that this is slightly different than #2807 since that is when the ball is outside the defence area, but inside the inflated _friendly_ defence area. In this case, the ball is outside both of those areas, so it would makes sense for the crease defenders to go and get the ball/clear it from friendly half. 
This seems somewhat unrelated to your current PR, but I think it's something we should definitely look into prior to Robocup. @itsarune what do you think?
(Also lmk when you're ready for a re-review)

**Comment by williamckha** [comment]
> I see. Though I feel like that this is slightly different than #2807 since that is when the ball is outside the defence area, but inside the inflated _friendly_ defence area. In this case, the ball is outside both of those areas, so it would makes sense for the crease defenders to go and get the ball/clear it from friendly half.

I think you're right, looks like that issue isn't really related then. From `getTeamWithPossession` I logged the position of a mostly stationary crease defender and it seems to jump back and forth between two different locations that are quite far apart. Might be something wrong with the teams being passed in and bot ids not being assigned consistently over time?
image

Since its position is always changing the possession tracker is unable to track the bot as near the ball for any meaningful amount of time, which is causing the issues with determining possession

**Comment by itsarune** [comment]
re: ball outside defense area and inside inflated defense area

I believe in this situation where it's awkward to allow the crease defender to violate its own motion constraint to kick away the ball and risk breaking the rule. I would be more comfortable with the goalie having some static ball checking and having auto kick on

File: src/software/ai/evaluation/defender_assignment_test.cpp [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+TEST_F(GroupGoalLanesByDensityTest, single_dense_cluster)
+{
+    std::vector<Point> threat_positions = {Point(-1, -1), Point(-1, -1.5), Point(-1, 0),
+                                           Point(-1, 2)};
+
+    // Create goal lanes from threat positions
+    std::vector<GoalLane> goal_lanes;
+    const Point FRIENDLY_GOAL_CENTER(-4.5, 0);
```
Comment (itsarune) [inline src/software/ai/evaluation/defender_assignment_test.cpp]: why not using the helper function from `field.h`

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                break;
+            case 1:
+                alignment = (event.common.world.ball().position().y() > 0)
+                                ? TbotsProto::CreaseDefenderAlignment::LEFT
+                                : TbotsProto::CreaseDefenderAlignment::RIGHT;
+                break;
+            case 2:
+                alignment = (event.common.world.ball().position().y() > 0)
+                                ? TbotsProto::CreaseDefenderAlignment::RIGHT
+                                : TbotsProto::CreaseDefenderAlignment::LEFT;
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: this style of assignment is not preferred

**Review by itsarune — COMMENTED** [review]
nice work, looks good! I have some minor code nits.

I also think that during attack, what do you think about having a `SupportTactic` during offense? It would provide a closer passing option near midfield when the wingers are blocked

**Comment by williamckha** [comment]
> nice work, looks good! I have some minor code nits.
> 
> I also think that during attack, what do you think about having a `SupportTactic` during offense? It would provide a closer passing option near midfield when the wingers are blocked

Yeah we could try out having a tactic like that. We could probably just take one of the crease defenders and move it up nearer to the halfway line when we have clear possession

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.h [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                                   setupOffensiveStrategy_A        = OffensiveState_S,
+            DefensiveState_S + Update_E / setupDefensiveStrategy_A = DefensiveState_S,
+            X + Update_E                                           = X);
+    }
+
+   private:
+    TbotsProto::AiConfig ai_config;
+    std::shared_ptr<ShootOrPassPlay> shoot_or_pass_play;
+    std::shared_ptr<DefensePlay> defense_play;
+    std::shared_ptr<PossessionTracker> possession_tracker;
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/offense/offense_play_fsm.h]: sensor_fusion has some possession logic, let's use the new possession tracker there instead?

https://github.com/UBC-Thunderbots/Software/blob/master/src/software/sensor_fusion/sensor_fusion.cpp

**Review by itsarune — COMMENTED** [review]
minor thing I remembered

File: src/software/ai/evaluation/defender_assignment.cpp (hunk 11 above)
Comment (williamckha) [reply src/software/ai/evaluation/defender_assignment.cpp]: done

File: src/software/ai/evaluation/defender_assignment.cpp (hunk 12 above)
Comment (williamckha) [reply src/software/ai/evaluation/defender_assignment.cpp]: done

File: src/software/math/math_functions.h (hunk 14 above)
Comment (williamckha) [reply src/software/math/math_functions.h]: done

File: src/software/ai/evaluation/defender_assignment_test.cpp (hunk 15 above)
Comment (williamckha) [reply src/software/ai/evaluation/defender_assignment_test.cpp]: changed to use field function

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 16 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: done

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.h (hunk 17 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.h]: I moved the possession tracker over to sensor fusion. sensor fusion now depends on `software/ai/evaluation/possession` since the possession tracker calls `getRobotWithEffectiveBallPossession`, wondering if this is ok or are we strict about coupling?

**Comment by williamckha** [comment]
Ready for review, I can open up new issues for using the breakbeam and having different levels of possession since this pr is getting pretty big

**Comment by nimazareian** [comment]
I've tried running AI vs AI and I got a SIGABRT 2 times (Had to run it for a couple minutes each time for it to happen). Have you seen this before?

Error message:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Cannot create a Line with two equal Points
```

The scenario that the robots are in once the SIGABRT happens:
image

Stack in CLion after the SIGABRT:
image

**Comment by williamckha** [comment]
> I've tried running AI vs AI and I got a SIGABRT 2 times (Had to run it for a couple minutes each time for it to happen). Have you seen this before?
> 
> Error message:
> 
> ```
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Cannot create a Line with two equal Points
> ```
> 
> The scenario that the robots are in once the SIGABRT happens: image
> 
> Stack in CLion after the SIGABRT: image

haven't seen this before, any specific conditions with where the ball is when this happens?

**Comment by nimazareian** [comment]
> haven't seen this before, any specific conditions with where the ball is when this happens?

No not sure. It happened once, and then the second time I ran the Blue Fullsystem I ran it in debug mode and after a couple of minutes of gameplay it randomly happened again.

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.h (hunk 17 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.h]: imo this coupling is okay as the evaluation folder is more of a utility class

File: src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp [hunk 18]
```diff
@@ -82,16 +83,23 @@ Rectangle GoalieFSM::getNoChipRectangle(const Field &field)
         field.friendlyGoalpostPos() + Vector(2 * ROBOT_MAX_RADIUS_METERS, 0));
 }
 
-Point GoalieFSM::findGoodChipTarget(const World &world)
+Point GoalieFSM::findGoodChipTarget(const World &world,
+                                    TbotsProto::GoalieTacticConfig goalie_tactic_config)
```
Comment (nimazareian) [inline src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp]: nit
```suggestion
                                    const TbotsProto::GoalieTacticConfig &goalie_tactic_config)
```

**Review by nimazareian — COMMENTED** [review]
very minor nit, but looks good! Good job finding the bug 👍

**Comment by nimazareian** [comment]
Also pull master so your CI passes

**Review by nimazareian — APPROVED** [review]

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        auto defenders_with_target_count = std::count_if(
+            crease_defender_assignments.begin(), crease_defender_assignments.begin() + i,
+            [&target](const auto& assignment) { return assignment.target == target; });
+
+        // Pick alignment based on how many crease defenders are already assigned to the
+        // target
+        auto alignment = TbotsProto::CreaseDefenderAlignment::CENTRE;
+        if (defenders_with_target_count == 1)
+        {
+            if (event.common.world.ball().position().y() > 0)
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:83]: Why is the logic different here based on number of assignments assigned to a tactic?

Comment (williamckha) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:83]: We want the first crease defender assigned to a target enemy to be centered on the target. Then we want the next two crease defenders (if any) to align themselves to the left and right of the first crease defender. Depending on where the ball is, we prioritize aligning the second crease defender either left or right to block more of the goal. I haven't really done any serious testing to see if this works, but based on the eye test it looks like it blocks more shots aimed at the near post

**Review by itsarune — APPROVED** [review]
LGTM!
